pkg/backends/calico/client.go

```diff
@@ -811,6 +885,258 @@ func (c *client) onUpdates(updates []api.Update) {
+		// We wrap each Service external IP in a ServiceExternalIPBlock struct to
+		// achieve the desired API structure, unpack that.
+		ipBlocks := v3res.Spec.ServiceExternalIPs
+		if len(ipBlocks) != 0 {
+			ipCidrs := make([]string, len(ipBlocks))
+			for i, ipBlock := range ipBlocks {
+				ipCidrs[i] = ipBlock.CIDR
+			}
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, strings.Join(ipCidrs, ",")))
+		} else {
+			nilValue, err := json.Marshal(nil)
+			if err != nil {
+				log.Errorf("Ignoring update: unable to serialize value %v: %v", v3res.Spec.ServiceExternalIPs, err)
+				return
+			}
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, string(nilValue)))
+		}
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+func (c *client) getServiceClusterIPsKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("svc_cluster_ips", key)
+
+	if v3res != nil && v3res.Spec.ServiceClusterIPs != nil {
+		ipBlocks := v3res.Spec.ServiceClusterIPs
+		// We wrap each Service Cluster IP in a ServiceClusterIPBlock to
+		// achieve the desired API structure. This unpacks that.
+		if len(ipBlocks) != 0 {
+			ipCidrs := make([]string, len(ipBlocks))
+			for i, ipBlock := range ipBlocks {
+				ipCidrs[i] = ipBlock.CIDR
+			}
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, strings.Join(ipCidrs, ",")))
+		} else {
+			nilValue, err := json.Marshal(nil)
+			if err != nil {
+				log.Errorf("Ignoring update: unable to serialize value %v: %v", v3res.Spec.ServiceClusterIPs, err)
+				return
+			}
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, string(nilValue)))
+		}
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+func (c *client) getNodeToNodeMeshKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("node_mesh", key)
+
+	if v3res != nil && v3res.Spec.NodeToNodeMeshEnabled != nil {
+		enabled := *v3res.Spec.NodeToNodeMeshEnabled
+		var val = nodeToNodeMeshEnabled
+		if !enabled {
+			val = nodeToNodeMeshDisabled
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, val))
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+func (c *client) getLogSeverityKVPair(v3res *apiv3.BGPConfiguration, key interface{}) {
+	globalKey := getBGPConfigKey("loglevel", key)
+
+	if v3res != nil && v3res.Spec.LogSeverityScreen != "" {
+		// Bird log level currently only supports granularity of none, debug and info.  Debug/Info are
+		// left unchanged, all others treated as none.
+		l := strings.ToLower(v3res.Spec.LogSeverityScreen)
+		switch l {
+		case "debug", "info":
+		default:
+			l = "none"
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(globalKey, l))
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(globalKey))
+	}
+}
+
+// returns the version for cidr, or 0 if the IP is not valid.
+func getVersion(cidr string) int {
```

Comment:
I wonder if we need this.  In a lot of existing places we say `strings.Contains(cidr, ":")` to identify IPv6, and AFAIK that is 100% robust.